File: java-client/src/main/java/com/couchbase/client/java/manager/query/AsyncCollectionQueryIndexManager.java
/*
 * Copyright 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.manager.query;

import com.couchbase.client.core.CoreKeyspace;
import com.couchbase.client.core.annotation.Stability;
import com.couchbase.client.core.api.query.CoreQueryOps;
import com.couchbase.client.core.cnc.RequestTracer;
import com.couchbase.client.core.error.CouchbaseException;
import com.couchbase.client.core.error.IndexExistsException;
import com.couchbase.client.core.error.IndexFailureException;
import com.couchbase.client.core.error.IndexNotFoundException;
import com.couchbase.client.core.manager.CoreCollectionQueryIndexManager;

import java.time.Duration;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static com.couchbase.client.java.manager.query.BuildQueryIndexOptions.buildDeferredQueryIndexesOptions;
import static com.couchbase.client.java.manager.query.CreatePrimaryQueryIndexOptions.createPrimaryQueryIndexOptions;
import static com.couchbase.client.java.manager.query.CreateQueryIndexOptions.createQueryIndexOptions;
import static com.couchbase.client.java.manager.query.DropPrimaryQueryIndexOptions.dropPrimaryQueryIndexOptions;
import static com.couchbase.client.java.manager.query.DropQueryIndexOptions.dropQueryIndexOptions;
import static com.couchbase.client.java.manager.query.GetAllQueryIndexesOptions.getAllQueryIndexesOptions;
import static com.couchbase.client.java.manager.query.WatchQueryIndexesOptions.watchQueryIndexesOptions;

/**
 * Performs management operations on query indexes at the Collection level.
 */
@Stability.Volatile
public class AsyncCollectionQueryIndexManager {

  private final CoreCollectionQueryIndexManager internal;

  /**
   * Creates a new {@link AsyncCollectionQueryIndexManager}.
   * <p>
   * This API is not intended to be called by the user directly, use {@link com.couchbase.client.java.AsyncCollection#queryIndexes()}
   * instead.
   */
  @Stability.Internal
  public AsyncCollectionQueryIndexManager(
    final CoreQueryOps queryOps,
    final RequestTracer requestTracer,
    final CoreKeyspace collection
  ) {
    this.internal = new CoreCollectionQueryIndexManager(queryOps, requestTracer, collection);
  }

  /**
   * Creates a named query index on this collection.
   *
   * @param indexName the name of the query index.
   * @param fields the collection of fields that are part of the index.
   * @throws IndexFailureException if creating the index failed (see reason for details).
   * @throws IndexExistsException if an index already exists with the given name on this collection.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createIndex(final String indexName,
                                             final Collection<String> fields) {
    return createIndex(indexName, fields, createQueryIndexOptions());
  }

  /**
   * Creates a named query index with custom options, on this collection.
   *
   * @param indexName the name of the query index.
   * @param fields the collection of fields that are part of the index.
   * @param options the custom options to apply.
   * @throws IndexFailureException if creating the index failed (see reason for details).
   * @throws IndexExistsException if an index already exists with the given name on this collection.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createIndex(final String indexName,
                                             final Collection<String> fields,
                                             final CreateQueryIndexOptions options) {
    return internal.createIndex(indexName, fields, options.build());
  }

  /**
   * Creates a primary query index on this collection.
   *
   * @throws IndexFailureException if creating the index failed (see reason for details).
   * @throws IndexExistsException if an index already exists with the given name on this collection.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createPrimaryIndex() {
    return createPrimaryIndex(createPrimaryQueryIndexOptions());
  }

  /**
   * Creates a primary query index with custom options, on this collection.
   *
   * @param options the custom options to apply.
   * @throws IndexFailureException if creating the index failed (see reason for details).
   * @throws IndexExistsException if an index already exists with the given name on this collection.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createPrimaryIndex(final CreatePrimaryQueryIndexOptions options) {
    return internal.createPrimaryIndex(options.build());
  }

  /**
   * Fetches all indexes on this collection.
   *
   * @return a {@link CompletableFuture} completing with a list of (potentially empty) indexes or failed with an error.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<List<QueryIndex>> getAllIndexes() {
    return getAllIndexes(getAllQueryIndexesOptions());
  }

  /**
   * Fetches all indexes from this collection with custom options.
   *
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing with a list of (potentially empty) indexes or failed with an error.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<List<QueryIndex>> getAllIndexes(final GetAllQueryIndexesOptions options) {
    return internal.getAllIndexes(options.build())
            .thenApply(rows -> rows.stream()
                    .map(row -> new QueryIndex(row))
                    .collect(Collectors.toList()));
  }

  /**
   * Drops the primary index from this collection.
   *
   * @throws IndexNotFoundException if the index does not exist.
   * @throws IndexFailureException if dropping the index failed (see reason for details).
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropPrimaryIndex() {
    return dropPrimaryIndex(dropPrimaryQueryIndexOptions());
  }

  /**
   * Drops the primary index from this collection with custom options.
   *
   * @param options the custom options to apply.
   * @throws IndexNotFoundException if the index does not exist.
   * @throws IndexFailureException if dropping the index failed (see reason for details).
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropPrimaryIndex(final DropPrimaryQueryIndexOptions options) {
    return internal.dropPrimaryIndex(options.build());
  }

  /**
   * Drops a query index from this collection.
   *
   * @param indexName the name of the index to drop.
   * @throws IndexNotFoundException if the index does not exist.
   * @throws IndexFailureException if dropping the index failed (see reason for details).
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropIndex(final String indexName) {
    return dropIndex(indexName, dropQueryIndexOptions());
  }

  /**
   * Drops a query index from this collection with custom options.
   *
   * @param indexName the name of the index to drop.
   * @param options the custom options to apply.
   * @throws IndexNotFoundException if the index does not exist.
   * @throws IndexFailureException if dropping the index failed (see reason for details).
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropIndex(final String indexName,
                                           final DropQueryIndexOptions options) {
    return internal.dropIndex(indexName, options.build());
  }

  /**
   * Builds all currently deferred indexes on this collection.
   *
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> buildDeferredIndexes() {
    return buildDeferredIndexes(buildDeferredQueryIndexesOptions());
  }

  /**
   * Builds all currently deferred indexes on this collection, with custom options.
   *
   * @param options the custom options to apply.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> buildDeferredIndexes(final BuildQueryIndexOptions options) {
    return internal.buildDeferredIndexes(options.build());
  }

  /**
   * Watches/Polls indexes on this collection until they are online.
   *
   * @param indexNames the names of the indexes to watch.
   * @param timeout the maximum amount of time the indexes should be watched.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> watchIndexes(final Collection<String> indexNames,
                                              final Duration timeout) {
    return watchIndexes(indexNames, timeout, watchQueryIndexesOptions());
  }

  /**
   * Watches/Polls indexes on this collection until they are online with custom options.
   *
   * @param indexNames the names of the indexes to watch.
   * @param timeout the maximum amount of time the indexes should be watched.
   * @param options the custom options to apply.
   * @throws CouchbaseException if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> watchIndexes(final Collection<String> indexNames,
                                              final Duration timeout, final WatchQueryIndexesOptions options) {
    return internal.watchIndexes(indexNames, timeout, options.build());
  }
}

File: scala-client/src/main/scala/com/couchbase/client/scala/manager/query/AsyncCollectionQueryIndexManager.scala
/*
 * Copyright (c) 2023 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.scala.manager.query

import com.couchbase.client.core.{Core, CoreKeyspace}
import com.couchbase.client.core.api.manager._
import com.couchbase.client.core.endpoint.http.CoreCommonOptions
import com.couchbase.client.core.manager.CoreCollectionQueryIndexManager
import com.couchbase.client.core.retry.{BestEffortRetryStrategy, RetryStrategy}
import com.couchbase.client.scala.AsyncCollection
import com.couchbase.client.scala.util.DurationConversions._
import com.couchbase.client.scala.util.FutureConversions

import java.{lang, util}
import scala.compat.java8.OptionConverters._
import scala.concurrent.duration.Duration
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._

/** Allows query indexes to be managed on this collection.
  *
  * @define Timeout        when the operation will timeout.  This will default to `timeoutConfig().managementTimeout
  *                        ()` in the
  *                        provided [[com.couchbase.client.scala.env.ClusterEnvironment]].
  * @define RetryStrategy  provides some control over how the SDK handles failures.  Will default to `retryStrategy()`
  *                        in the provided [[com.couchbase.client.scala.env.ClusterEnvironment]].
  */
class AsyncCollectionQueryIndexManager(
    private[scala] val collection: AsyncCollection,
    private val keyspace: CoreKeyspace
)(
    implicit val ec: ExecutionContext
) {
  private[scala] val internal = new CoreCollectionQueryIndexManager(
    collection.couchbaseOps.queryOps(),
    collection.core.coreResources.requestTracer,
    keyspace
  )
  private[scala] val DefaultTimeout: Duration =
    collection.couchbaseOps.environment.timeoutConfig.managementTimeout
  private[scala] val DefaultRetryStrategy = collection.couchbaseOps.environment.retryStrategy

  /** Gets all indexes on this collection.
    *
    * @param timeout       $Timeout
    * @param retryStrategy $RetryStrategy
    */
  def getAllIndexes(
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Seq[QueryIndex]] = {
    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.getAllIndexes(
          new CoreGetAllQueryIndexesOptions {
            override def scopeName(): String = null

            override def collectionName(): String = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(
        result =>
          // toSeq required for Scala 2.13
          result.asScala.toSeq.map(
            i =>
              QueryIndex(
                i.name,
                Some(i.primary),
                Option(i.`type`),
                i.state,
                i.keyspace,
                i.indexKey.asScala.toSeq.map(v => v.textValue),
                i.condition.asScala,
                i.partition.asScala,
                Option(i.bucketName),
                i.scopeName.asScala
              )
          )
      )
  }

  /** Creates a new query index on this collection, with the specified parameters.
    *
    * @param indexName      the name of the index.
    * @param ignoreIfExists if an index with the same name already exists, the operation will fail.
    * @param numReplicas    how many replicas of the index to create.
    * @param deferred       set to true to defer building the index until [[buildDeferredIndexes]] is called.  This can
    *                       provide improved performance when creating multiple indexes.
    * @param timeout        $Timeout
    * @param retryStrategy  $RetryStrategy
    */
  def createIndex(
      indexName: String,
      fields: Iterable[String],
      ignoreIfExists: Boolean = false,
      numReplicas: Option[Int] = None,
      deferred: Option[Boolean] = None,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Unit] = {
    val _deferred       = deferred
    val _numReplicas    = numReplicas
    val _ignoreIfExists = ignoreIfExists

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.createIndex(
          indexName,
          fields.toSeq.asJava,
          new CoreCreateQueryIndexOptions {
            override def ignoreIfExists(): Boolean = _ignoreIfExists

            override def `with`(): util.Map[String, AnyRef] = null

            override def scopeAndCollection(): CoreScopeAndCollection = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)

            override def numReplicas(): Integer = _numReplicas.map(Integer.valueOf).orNull

            override def deferred(): lang.Boolean =
              _deferred.map(v => lang.Boolean.valueOf(v)).orNull
          }
        )
      )
      .map(_ => ())
  }

  /** Creates a new primary query index on this collection, with the specified parameters.
    *
    * @param indexName      the name of the index.  If not set the server assigns the default primary index name.
    * @param ignoreIfExists if a primary index already exists, the operation will fail.
    * @param numReplicas    how many replicas of the index to create.
    * @param deferred       set to true to defer building the index until [[buildDeferredIndexes]] is called.  This can
    *                       provide improved performance when creating multiple indexes.
    * @param timeout        $Timeout
    * @param retryStrategy  $RetryStrategy
    */
  def createPrimaryIndex(
      indexName: Option[String] = None,
      ignoreIfExists: Boolean = false,
      numReplicas: Option[Int] = None,
      deferred: Option[Boolean] = None,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Unit] = {
    val _deferred       = deferred
    val _numReplicas    = numReplicas
    val _ignoreIfExists = ignoreIfExists
    val _indexName      = indexName

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.createPrimaryIndex(
          new CoreCreatePrimaryQueryIndexOptions {
            override def ignoreIfExists(): Boolean = _ignoreIfExists

            override def `with`(): util.Map[String, AnyRef] = null

            override def scopeAndCollection(): CoreScopeAndCollection = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)

            override def numReplicas(): Integer = _numReplicas.map(Integer.valueOf).orNull

            override def deferred(): lang.Boolean =
              _deferred.map(v => lang.Boolean.valueOf(v)).orNull

            override def indexName(): String = _indexName.orNull
          }
        )
      )
      .map(_ => ())
  }

  /** Drops an existing index on this collection.
    *
    * @param indexName         the name of the index.
    * @param ignoreIfNotExists sets whether the operation should fail if the index does not exists
    * @param timeout           $Timeout
    * @param retryStrategy     $RetryStrategy
    */
  def dropIndex(
      indexName: String,
      ignoreIfNotExists: Boolean = false,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Unit] = {
    val _ignoreIfNotExists = ignoreIfNotExists

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.dropIndex(
          indexName,
          new CoreDropQueryIndexOptions {
            override def ignoreIfNotExists(): Boolean = _ignoreIfNotExists

            override def scopeAndCollection(): CoreScopeAndCollection = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  /** Drops an existing primary index on this collection.
    *
    * @param ignoreIfNotExists sets whether the operation should fail if the index does not exists
    * @param timeout           $Timeout
    * @param retryStrategy     $RetryStrategy
    */
  def dropPrimaryIndex(
      ignoreIfNotExists: Boolean = false,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Unit] = {
    val _ignoreIfNotExists = ignoreIfNotExists

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.dropPrimaryIndex(
          new CoreDropPrimaryQueryIndexOptions {
            override def ignoreIfNotExists(): Boolean = _ignoreIfNotExists

            override def scopeAndCollection(): CoreScopeAndCollection = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  /** Polls the specified indexes on this collection until they are all online.
    *
    * @param indexNames    the indexes to poll.
    * @param watchPrimary  include the bucket's primary index.  If the bucket has no primary index, the operation
    *                      will fail with `IndexNotFoundException`
    * @param timeout       when the operation will timeout.
    * @param retryStrategy $RetryStrategy
    */
  def watchIndexes(
      indexNames: Iterable[String],
      timeout: Duration,
      watchPrimary: Boolean = false,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Unit] = {
    val _watchPrimary = watchPrimary

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.watchIndexes(
          indexNames.toSeq.asJava,
          timeout,
          new CoreWatchQueryIndexesOptions {
            override def watchPrimary(): Boolean = _watchPrimary

            override def scopeAndCollection(): CoreScopeAndCollection = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  /** Build all deferred indexes on this collection.
    *
    * @param timeout        $Timeout
    * @param retryStrategy  $RetryStrategy
    */
  def buildDeferredIndexes(
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): Future[Unit] = {
    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.buildDeferredIndexes(
          new CoreBuildQueryIndexOptions {
            override def scopeAndCollection(): CoreScopeAndCollection = null

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  def makeCommonOptions(
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): CoreCommonOptions = {
    CoreCommonOptions.of(
      if (timeout == DefaultTimeout) null else timeout,
      if (retryStrategy == DefaultRetryStrategy) null else retryStrategy,
      null
    )
  }
}
